Try removing default align-items styles from flex layout

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We're increasingly running into issues caused by the `align-items` property that gets added by default to flex layouts. In horizontal layouts, we're adding `align-items: center` and in vertical ones `align-items: flex-start`. For most blocks, it's not useful to set these properties to these particular values, and in some cases it's actively working against their layout potential. 

I've only found one case where setting this value by default is needed: the Social Icons block in its vertical orientation, as described in #38069.

#46133 and #45117 describe layout scenarios that should be fairly straightforward to build and become difficult or impossible with the current `align-items` defaults.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* This PR tries removing `align-items` from the default flex layout. 
* Given the special case of Social Icons, it adds the property back in this block's styles.
* It also removes a fallback style from the Gallery block which should no longer be needed.

It's still possible to manipulate the value to some extent in horizontal orientations (such as Row block) with the block alignment settings, and the justification controls for vertical orientations (such as Stack block).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Add the following blocks to a page: Social Icons, Buttons, Gallery, Navigation, Row, Stack.
2. Add some contents to each and check that they render as expected in the editor and the front end.
3. Check that the layouts described in #46133 and #45117 are now possible to build with no custom CSS additions.

 
### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

A Stack inside a Row should now by default take up the full height of the tallest block inside the Row:

<img width="753" alt="Screenshot 2023-01-12 at 3 42 05 pm" src="https://user-images.githubusercontent.com/8096000/211977945-5964152f-471a-4952-95ba-f357e5c64f3c.png">

tellthemachines committed:
Try removing default align-items styles from flex layout

tellthemachines committed:
Try removing gallery block fallback.

andrewserong commented:
Oh, what an interesting problem, thanks for diving in! At first glance, I'm not sure how feasible it'll be to remove some of the default values, because the Stack block launched with the idea that by default blocks have the width of their content (a bit like the social icons block). Here's what I noticed on an existing Stack block:

| Trunk | This PR |
| --- | --- |
| <img width="700" src="https://user-images.githubusercontent.com/14988353/212210587-c7332e9a-1a52-4d5b-be95-235ca645089d.png" /> | <img width="700" alt="image" src="https://user-images.githubusercontent.com/14988353/212210610-90cfcb87-7556-468c-bf94-1f1d96915cbe.png"> |

Now that the layout features have been out in the wild for a while, if we do remove things from the default values, I think we'll need to be quite careful to see what sorts of existing patterns might have been depending on these values 🤔 

> A Stack inside a Row should now by default take up the full height of the tallest block inside the Row:

Rather than changing the defaults, is there a way that we could add a control that would support stretching children to fill the container, and treat this an enhancement?

andrewserong commented:
Now that https://github.com/WordPress/gutenberg/pull/47134 has landed, shall we close out this one? Or do you think it's still worth pursuing changing the default options 🤔

tellthemachines commented:
I think it's still good to remove defaults that aren't useful, just need to work out if it won't break anything. Ideally those styles would never have been added in the first place 😅

andrewserong commented:
Okie-doke!

>  Ideally those styles would never have been added in the first place 😅

Ah, the perennial backwards compatibility challenge! 😄

andrewserong commented:
I was curious: TIL it turns out the `align-items: center` rule dates back to the initial PR that introduced flex (https://github.com/WordPress/gutenberg/pull/33359) (just in case that helps with any archaeological digs into the layout support 🙂)

mehne commented:
I can second this. 
I reported an in-depth description why this is problematic here too: #49502

ramonjd commented:
I'm not terribly familiar with the context, but I've been testing this and can confirm that the default `align-items` CSS property is not present by default.

Inserting a group block (Stack) in trunk, I get:

<img width="274" alt="Screenshot 2023-04-21 at 3 20 47 pm" src="https://user-images.githubusercontent.com/6458278/233548100-76ef0af0-b69c-4445-95b7-91305b668f2a.png">


In this PR I get:
<img width="263" alt="Screenshot 2023-04-21 at 3 25 54 pm" src="https://user-images.githubusercontent.com/6458278/233548108-f0ead8c7-eaa2-47a6-8522-95dc319ca1a8.png">


Playing around with justification controls I see that `align-items` is correctly added.

The social links block has the following styles too:

```css
.wp-block-social-links.is-vertical {
    align-items: flex-start;
}
```

All looks good on the frontend for me. 

Are there some other patterns that need testing?

mehne commented:
> I'm not terribly familiar with the context, but I've been testing this and can confirm that the default `align-items` CSS property is not present by default.
> 
> Inserting a group block (Stack) in trunk, I get:
> 
> <img alt="Screenshot 2023-04-21 at 3 20 47 pm" width="274" src="https://user-images.githubusercontent.com/6458278/233548100-76ef0af0-b69c-4445-95b7-91305b668f2a.png">
> 
> In this PR I get: <img alt="Screenshot 2023-04-21 at 3 25 54 pm" width="263" src="https://user-images.githubusercontent.com/6458278/233548108-f0ead8c7-eaa2-47a6-8522-95dc319ca1a8.png">
> 
> Playing around with justification controls I see that `align-items` is correctly added.
> 
> The social links block has the following styles too:
> 
> ```css
> .wp-block-social-links.is-vertical {
>     align-items: flex-start;
> }
> ```
> 
> All looks good on the frontend for me.
> 
> Are there some other patterns that need testing?

Sorry, but "playing around with justification" is not whats the problem here. It is about the default value. And it is set, like the OP described and like I describeb in my linked issue. 

Individual Styles from blocks like the social links block you described is not the problem too. Please read the initial post carefully. It is about the default css "align-items: center" that is added but should not!
Like here:
![Screenshot 2023-04-21 081043](https://user-images.githubusercontent.com/7791468/233554139-e504915e-3e43-48d3-aa81-a616b3c4b2b3.png)

ramonjd commented:
> Please read the initial post carefully. It is about the default css "align-items: center" that is added but should not!

Would you be able to test this PR and offer your expertise? I think it'd help move this issue along if so, especially if there are things that I've missed after testing it for the first time.

Hence my question: 

> Are there some other patterns that need testing?

The PR description specifically mentions the following:

> In horizontal layouts, we're adding align-items: center and in vertical ones align-items: flex-start. For most blocks, it's not useful to set these properties to these particular values, and in some cases it's actively working against their layout potential.

tellthemachines commented:
> Are there some other patterns that need testing?

Thanks for taking a look! I've meaning to come back to this PR and do some more testing on it 😅 
I think mainly it's a matter of checking that flex layout blocks with their default settings (i.e. no changes to alignment/justification) still display the same on trunk and in this branch, to make sure there are no visual regressions.

Might be worth taking a look at the WP Directory to see if there are third party blocks using this layout type (probably) and check those, but I'm more than happy to do that myself.

mehne commented:
> The PR description specifically mentions the following:
> 
> > In horizontal layouts, we're adding align-items: center and in vertical ones align-items: flex-start. For most blocks, it's not useful to set these properties to these particular values, and in some cases it's actively working against their layout potential.

With "we are adding" the PR means Wordpress (Gutenberg) is adding those on default. See the Screenshot in my last post and the linked issue. When you add a block and don't change anything to the alignment, you get those classes. 

In my case in a custom block with "__experimental" activated. 
More details to my special case in this issue: #49502

ramonjd commented:
> With "we are adding" the PR means Wordpress (Gutenberg) is adding those on default. See the Screenshot in my last post and the linked issue. When you add a block and don't change anything to the alignment, you get those classes.

That's what I understood too, and formed the basis of my quick test. 

This PR is not marked as resolving your issue. If you think it should cover it, it might be more helpful to the PR author to test it and provide feedback as to where changes or improvements could be made.

tellthemachines committed:
Fix default control selection

tellthemachines committed:
Add default justification for social icons instead of CSS override

tellthemachines commented:
I rebased and updated the default alignment control values to the new defaults. I've also added a default `justifyContent` value to Social Icons block, instead of the previous CSS override.

In testing the core blocks, there are a few discrepancies with trunk in specific cases, namely:

* Buttons block where buttons have different heights
* Row block where items have different heights
* Stack block where items have background color or other visible indication of their width

Trunk:

<img width="631" alt="Screenshot 2023-04-24 at 1 59 53 pm" src="https://user-images.githubusercontent.com/8096000/233904034-61de9fff-4343-4f27-be7a-25eb18cd44ea.png">
<img width="907" alt="Screenshot 2023-04-24 at 2 00 01 pm" src="https://user-images.githubusercontent.com/8096000/233904070-43f49f63-7f77-47f4-a82f-77951a885013.png">
<img width="640" alt="Screenshot 2023-04-24 at 2 00 16 pm" src="https://user-images.githubusercontent.com/8096000/233904092-2277a9b8-0a61-42da-86b0-aa8626c77fc8.png">

This branch:

<img width="626" alt="Screenshot 2023-04-24 at 2 00 37 pm" src="https://user-images.githubusercontent.com/8096000/233904123-a1a2f0ed-89b5-40cb-a6fa-4c1b7eb0bf06.png">
<img width="1075" alt="Screenshot 2023-04-24 at 2 00 47 pm" src="https://user-images.githubusercontent.com/8096000/233904132-d39e55ec-d838-431a-a0e0-00cd65dbf313.png">
<img width="673" alt="Screenshot 2023-04-24 at 2 00 55 pm" src="https://user-images.githubusercontent.com/8096000/233904141-470b76de-c674-4ca7-b49a-25c9c25cf717.png">

For Row and Stack blocks at least, I don't believe we should add a default alignment value such as I added to Social Icons, as that would mostly defeat the purpose of this PR. We _want_ the default to change. However, this may bring about breaking changes to websites relying on the current defaults. We could conceivably work around this by adding a default alignment that matches the current one and then adding an override in the Row/Stack block variation settings to make all _new_ Row and Stack blocks have the new default. It's a bit messy, and in that scenario I think that programatically added blocks will still get the legacy defaults unless explicitly specified, but I can't think of any way to write a deprecation for this as there are no changes to the actual blocks, so that may be our only alternative to breaking changes.

For Buttons, adding a default center alignment might work well.

Further feedback and ideas welcome!

tellthemachines commented:
Closing as there's no obvious solution to this one that preserves back-compat, and default behaviour can be changed via the flex alignment controls. Thanks folks!
